File: server/scripts/compose_navy_frame.py
#!/usr/bin/env python3.11
"""
Navy Frame compositor for Lifework LinkedIn images.
Usage: python3.11 compose_navy_frame.py <photo_path> <tangram_path> <serif_font_path> <sans_font_path> <category_label> <output_path>

Canvas: 1080×1080px
- Outer frame: navy #1a2744, full bleed
- Inner photo: reduced to fit above a taller bottom rail
- Bottom rail: 200px tall — logo lockup extends to ~mid-image width, large category label on right
"""
import sys
from PIL import Image, ImageDraw, ImageFont

def hex_to_rgb(h):
    h = h.lstrip('#')
    return tuple(int(h[i:i+2], 16) for i in (0, 2, 4))

NAVY  = hex_to_rgb('#1a2744')
CREAM = hex_to_rgb('#f5f0e8')
GOLD  = hex_to_rgb('#d4a634')

CANVAS_W, CANVAS_H = 1080, 1080
MARGIN = 60

# Taller rail to accommodate larger branding
RAIL_H = 200
RAIL_Y = CANVAS_H - RAIL_H  # 880

# Photo fills the space above the rail (minus top/side margins)
PHOTO_W = CANVAS_W - 2 * MARGIN   # 960
PHOTO_H = RAIL_Y - MARGIN          # 820  (was 890 — slightly shorter to give rail room)

# Branding sizes — tangram + wordmark together reach roughly mid-canvas (~540px)
MARK_SIZE          = 90    # tangram square
WORDMARK_FONT_SIZE = 90    # "Lifework" serif — large enough to reach mid-image
LABEL_FONT_SIZE    = 38    # category label on right — prominent


def compose(photo_path, tangram_path, serif_font_path, sans_font_path, category_label, output_path):
    # 1. Navy canvas
    canvas = Image.new('RGB', (CANVAS_W, CANVAS_H), NAVY)

    # 2. Inner photo
    photo = Image.open(photo_path).convert('RGB')
    photo = photo.resize((PHOTO_W, PHOTO_H), Image.LANCZOS)
    canvas.paste(photo, (MARGIN, MARGIN))

    # 3. Tangram mark — vertically centred in rail
    tangram = Image.open(tangram_path).convert('RGBA')
    tangram = tangram.resize((MARK_SIZE, MARK_SIZE), Image.LANCZOS)
    mark_x = MARGIN
    mark_y = RAIL_Y + (RAIL_H - MARK_SIZE) // 2
    canvas.paste(tangram, (mark_x, mark_y), tangram)

    # 4. "Lifework" wordmark — serif, cream, vertically centred in rail
    draw = ImageDraw.Draw(canvas)
    try:
        serif_font = ImageFont.truetype(serif_font_path, size=WORDMARK_FONT_SIZE)
    except Exception:
        serif_font = ImageFont.load_default()

    wordmark_text = "Lifework"
    wordmark_x = mark_x + MARK_SIZE + 18
    bbox = draw.textbbox((0, 0), wordmark_text, font=serif_font)
    text_h = bbox[3] - bbox[1]
    wordmark_y = RAIL_Y + (RAIL_H - text_h) // 2 - bbox[1]
    draw.text((wordmark_x, wordmark_y), wordmark_text, font=serif_font, fill=CREAM)

    # 5. Category label — uppercase sans, right-aligned, 78% opacity
    try:
        sans_font = ImageFont.truetype(sans_font_path, size=LABEL_FONT_SIZE)
    except Exception:
        sans_font = ImageFont.load_default()

    label_text = category_label.upper()
    label_bbox = draw.textbbox((0, 0), label_text, font=sans_font)
    label_w = label_bbox[2] - label_bbox[0]
    label_h = label_bbox[3] - label_bbox[1]
    label_x = CANVAS_W - MARGIN - label_w
    label_y = RAIL_Y + (RAIL_H - label_h) // 2 - label_bbox[1]

    opacity = 0.78
    blended = tuple(int(c * opacity + n * (1 - opacity)) for c, n in zip(CREAM, NAVY))
    draw.text((label_x, label_y), label_text, font=sans_font, fill=blended)

    # 6. Save
    canvas.save(output_path, 'PNG', optimize=True)
    print(f"Saved: {output_path}")


if __name__ == '__main__':
    if len(sys.argv) != 7:
        print("Usage: compose_navy_frame.py <photo> <tangram> <serif_font> <sans_font> <label> <output>")
        sys.exit(1)
    compose(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], sys.argv[6])

File: server/scripts/composeNavyFrame.ts
/**
 * Navy Frame compositor — pure Node.js/sharp implementation.
 *
 * Replaces the Python/PIL compositor so the image pipeline works in the
 * deployed production container (which does not have Python or Pillow).
 *
 * Canvas: 1080×1080px
 * - Outer frame: navy #1a2744, full bleed
 * - Inner photo: 960×820px, centred horizontally (60px each side), 30px from top
 * - Bottom rail: 200px, navy, full width
 * - Tangram mark: 90×90px, left edge of rail (x=60), vertically centred
 * - Lifework wordmark: NotoSerif, ~72px tall, cream #f5f0e8, after tangram
 * - Category label: NotoSans, ~32px tall, cream, letter-spaced, right-aligned
 *
 * Text is rendered via sharp's native Pango text input (not SVG/librsvg),
 * which requires a fontfile path on disk. Fonts are fetched from CDN on
 * first use, written to a process-lifetime temp file, and reused thereafter.
 * A local filesystem fallback is used when CDN is unavailable.
 */

import sharp from "sharp";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import * as https from "https";
import * as http from "http";

// CDN URLs for fonts (uploaded as static webdev assets — never expire)
const SERIF_FONT_URL =
  "https://d2xsxph8kpxj0f.cloudfront.net/107696804/kFbbE6kqNApXGDFpQJUGV7/NotoSerif-Regular_060c2d57.ttf";
const SANS_FONT_URL =
  "https://d2xsxph8kpxj0f.cloudfront.net/107696804/kFbbE6kqNApXGDFpQJUGV7/NotoSans-Regular_b64a5eb9.ttf";
const TANGRAM_URL =
  "https://d2xsxph8kpxj0f.cloudfront.net/107696804/kFbbE6kqNApXGDFpQJUGV7/tangram_ec333843.png";

// Local fallback paths (available in sandbox/dev, may not exist in production)
const LOCAL_SERIF = "/usr/share/fonts/truetype/noto/NotoSerif-Regular.ttf";
const LOCAL_SANS = "/usr/share/fonts/truetype/noto/NotoSans-Regular.ttf";
const LOCAL_TANGRAM = "/home/ubuntu/webdev-static-assets/tangram.png";

// Process-lifetime temp file paths for fonts (written once, reused)
let serifTmpPath: string | null = null;
let sansTmpPath: string | null = null;

// In-memory buffer cache for the tangram PNG
let tangramBuffer: Buffer | null = null;

function fetchBuffer(url: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const client = url.startsWith("https") ? https : http;
    client
      .get(url, (res) => {
        if (res.statusCode !== 200) {
          reject(new Error(`Failed to fetch ${url}: HTTP ${res.statusCode}`));
          return;
        }
        const chunks: Buffer[] = [];
        res.on("data", (chunk: Buffer) => chunks.push(chunk));
        res.on("end", () => resolve(Buffer.concat(chunks)));
        res.on("error", reject);
      })
      .on("error", reject);
  });
}

async function loadFontToTmpFile(
  cdnUrl: string,
  localPath: string,
  suffix: string
): Promise<string> {
  const tmpPath = path.join(os.tmpdir(), `lw_font_${suffix}_${process.pid}.ttf`);
  if (fs.existsSync(tmpPath)) return tmpPath;

  let buf: Buffer;
  if (fs.existsSync(localPath)) {
    buf = fs.readFileSync(localPath);
  } else {
    buf = await fetchBuffer(cdnUrl);
  }
  fs.writeFileSync(tmpPath, buf);
  return tmpPath;
}

async function loadTangram(): Promise<Buffer> {
  if (tangramBuffer) return tangramBuffer;
  if (fs.existsSync(LOCAL_TANGRAM)) {
    tangramBuffer = fs.readFileSync(LOCAL_TANGRAM);
  } else {
    tangramBuffer = await fetchBuffer(TANGRAM_URL);
  }
  return tangramBuffer;
}

async function ensureFonts(): Promise<{ serifPath: string; sansPath: string }> {
  if (!serifTmpPath) {
    serifTmpPath = await loadFontToTmpFile(SERIF_FONT_URL, LOCAL_SERIF, "serif");
  }
  if (!sansTmpPath) {
    sansTmpPath = await loadFontToTmpFile(SANS_FONT_URL, LOCAL_SANS, "sans");
  }
  return { serifPath: serifTmpPath, sansPath: sansTmpPath };
}

/**
 * Compose a 1080×1080 Navy Frame image.
 *
 * @param photoBuffer  Raw bytes of the inner photograph (any format sharp accepts)
 * @param categoryLabel  e.g. "STRENGTHS", "REFLECTIONS", "IN PRACTICE", "LIFELINE"
 * @returns  PNG buffer of the composited 1080×1080 image
 */
export async function composeNavyFrame(
  photoBuffer: Buffer,
  categoryLabel: string
): Promise<Buffer> {
  // Canvas constants
  const W = 1080;
  const H = 1080;
  const RAIL_H = 200;
  const PHOTO_W = 960;
  const PHOTO_H = 820;
  const PHOTO_X = Math.round((W - PHOTO_W) / 2); // 60
  const PHOTO_Y = 30;
  const TANGRAM_SIZE = 90;
  const TANGRAM_X = 60;
  const TANGRAM_Y = Math.round(H - RAIL_H + (RAIL_H - TANGRAM_SIZE) / 2);

  // Rail vertical centre
  const railCenterY = H - RAIL_H + Math.round(RAIL_H / 2);

  // Load fonts and tangram
  const { serifPath, sansPath } = await ensureFonts();
  const tgramBuf = await loadTangram();

  // ── Wordmark: "Lifework" in NotoSerif, ~72px tall ──────────────────────────
  const WORDMARK_H = 72;
  const wordmarkRaw = await sharp({
    text: {
      text: '<span foreground="#f5f0e8" font_family="NotoSerif">Lifework</span>',
      fontfile: serifPath,
      rgba: true,
      dpi: 300,
    },
  } as Parameters<typeof sharp>[0])
    .png()
    .toBuffer();
  const wMeta = await sharp(wordmarkRaw).metadata();
  const wordmarkW = Math.round(wMeta.width! * (WORDMARK_H / wMeta.height!));
  const wordmarkFinal = await sharp(wordmarkRaw)
    .resize(wordmarkW, WORDMARK_H)
    .png()
    .toBuffer();

  // ── Category label: NotoSans, ~32px tall, letter-spaced ────────────────────
  const LABEL_H = 32;
  const labelRaw = await sharp({
    text: {
      text: `<span foreground="#f5f0e8" letter_spacing="3000" font_family="NotoSans">${categoryLabel}</span>`,
      fontfile: sansPath,
      rgba: true,
      dpi: 300,
    },
  } as Parameters<typeof sharp>[0])
    .png()
    .toBuffer();
  const lMeta = await sharp(labelRaw).metadata();
  const labelW = Math.round(lMeta.width! * (LABEL_H / lMeta.height!));
  const labelFinal = await sharp(labelRaw)
    .resize(labelW, LABEL_H)
    .png()
    .toBuffer();

  // ── Tangram mark: 90×90px ───────────────────────────────────────────────────
  const tangramFinal = await sharp(tgramBuf)
    .resize(TANGRAM_SIZE, TANGRAM_SIZE)
    .png()
    .toBuffer();

  // ── Inner photo: 960×820px, cover crop ─────────────────────────────────────
  const photoFinal = await sharp(photoBuffer)
    .resize(PHOTO_W, PHOTO_H, { fit: "cover", position: "centre" })
    .png()
    .toBuffer();

  // ── Composite positions ─────────────────────────────────────────────────────
  const wordmarkTop = railCenterY - Math.round(WORDMARK_H / 2);
  const wordmarkLeft = TANGRAM_X + TANGRAM_SIZE + 24;
  const labelTop = railCenterY - Math.round(LABEL_H / 2);
  const labelLeft = W - 60 - labelW;

  // ── Final composite ─────────────────────────────────────────────────────────
  return sharp({
    create: {
      width: W,
      height: H,
      channels: 4,
      background: { r: 26, g: 39, b: 68, alpha: 1 },
    },
  })
    .composite([
      { input: photoFinal, top: PHOTO_Y, left: PHOTO_X },
      { input: tangramFinal, top: TANGRAM_Y, left: TANGRAM_X },
      { input: wordmarkFinal, top: wordmarkTop, left: wordmarkLeft },
      { input: labelFinal, top: labelTop, left: labelLeft },
    ])
    .png()
    .toBuffer();
}
